### ISSUE TITLE
Liquify and the future of this extension!

### ISSUE THREAD

**Issue by panoply** [panoply/vscode-liquid#56]
Hey all users 👋 

Firstly, thank you for using this extension. As some of you aware, over the last 12 months I have been working on the next version of this project with the end goal of bringing modern IDE feature to multiple variations of the Liquid Language. This endeavour that has taken a very long time to conclude and this mainly comes down to a few things, like the daily grind of adulthood, work, covid and mostly because Liquid is a customisable language with no real formal grammar, it's a templating language so thats its very nature. Coming up with a solution that will be able to accommodate its various use cases was rather difficult to say the least. Multiple rewrites, lack of sleep and a somewhat steep adoption in new techniques and approaches in software development/engineering etc has resulted in delayed releases.

This issue aims to provide any curious users awaiting on the next release a feature ship list as well as give some insight into what to expect. 

### Liquify
The next release will be shipping under a new name, which is "Liquify"  the reason for this is because support will also be made available in Sublime and Atom. Additionally, the project will open source a multitude of tools and third parties, like its AST parser, TextMate grammars, LSP client and various rollup bundle utilities which are used in the development environment that people may find useful in projects unrelated to the tool. 

### Variation Specific
Liquify will treat each Liquid Variation differently. Users who are developing with Jekyll do not require Shopify variation features, for example the `{% schema %}` tag  is unique to the Shopify Liquid variation only, it is not available in the Jekyll/Standard/Eleventy variation. The same logic applies to Jekyll/Eleventy where something like frontmatter is not supported in Shopify and therefore users should not be subject to features that are not supported in the development environment they are working within. 

In Liquify, you will define your variation on a per-project basis, this will tell the tool to load and execute according to the `engine` the developer has defined in their `.liquidrc` file.

### Language Server Protocol
Liquify will use the Language Server Protocol (LSP for short) to facilitate its features. LSP is generally leveraged and used by official languages like Python, Elixir, JavaScript and more, so Liquid will be hanging out with the big guys in the code game. More importantly, by leveraging LSP Liquify will become available to any editor that understands the protocol.

### Completions
Liquify will provide code completions of tags, objects and filters for multiple variations of Liquid, ie: Shopify, Jekyll, Eleventy etc. Completions will really empower your development workflow, especially for Shopify developers as the entire Shopify theme API will be at your fingertips. For those using Standard Liquid or Jekyll/Eleventy you were not forgotten and completions are facilitated also.

#### Scope Completions
It's common practice in Liquid to set a variable and then reference that variable or split it up and iterate over the values, whatever the case might be, if you work with this templating language, you're going to use `{% assign %}` or `{% capture %}`. Liquify will keep track of your variables, treating them as "scopes" in each file they are defined and from here will automatically be supplied to completions. 

#### Extendable Completions
Completions will be extendable, this means Jekyll/Eleventy developers will be able to provide completions to any custom Liquid tags/filters that you may be implementing through a plugin. You can simply define any extendable completions within the `.liquidrc` and boom, they are catapulted right into your document tag.

#### Externally Referenced Completions
Liquid is often referencing data from external files. In Shopify (for example) you will reference `locales`, or `settings_data.json` whereas in Jekyll or Eleventy you will reference `_collection` or `_data` yaml/json files. In most cases you call upon this data using Liquid and while that has worked just fine for the last decade, the landscape has changed and those files can be monolithic. Services like [DatoCMS](https://www.datocms.com/) empower developers by generating these files from a series of models and so it's not always easy remembering what is what. Liquify solves this issue and will scan your workspace for any files that the variation may require or use and from here supply them to you. It does this progressively and will cache results. 

> This feature may not make it to the official release, but rest assured will be available in a minor post 3.0.

#### Embedded Language Completions
The completion feature goes one step further by providing embedded language support for code blocks like the Shopify `{% schema %}` tag where the contained JSON will support completions (so no more shitty or otherwise annoying snippet patches, hooray!) – Embedded language completions use Language Services made available by vscode and you will able to assert new embedded language blocks to specific tags, truly customisable.

### Hover Descriptives
Liquify will provide users with hover descriptives of syntax, this means you will never have to goto a documentation page again. The tool will provide you with the official description of each object, object property, tag, or filter relative to the variation it belongs to. This feature will really help devs leverage the language to its fullest and use aspects and tags that you might otherwise not have known existed.  

### Validations/Linting
Liquify will support syntax validations and linting capabilities. This is a great feature that not only helps you write clean and concise Liquid code, but also will prevent you from writing incorrect syntax. Even though Liquid is a very basic and simple language, it can sometimes get quite chaotic in large code bases and scouring code for parse issues can get tiresome. This feature aims to prevent this and enables you or your team to implement a strict styleguide of how Liquid syntax should be expressed, while at the same automatically fixing errors that may occur. 

Users will have the option to control linting via the available rulesets, which can be customized to suit your desired tastes. Want auto-spacing within tags? No problem. Want enforce `{% capture %}` only variables, No problem. Whatever the case, you will get control via the rulesets. 

### Codelens / Document Linking
In variations outside of the Standard variation, we have `{% include %}` (or `{% render %}` in Shopify) tags which allow developers to reference an external file, otherwise known as snippets or partials. Liquify understands these type of tags and will provide a document link to the file that is being referenced, think of this like `import` in JavaScript. Simply, `cmd+click` to open the linked file or peek the content of the include via codelens. Its dope fam, really dope. 

### Formatting
A lot of developers use this extension for the formatting capabilities, you'll get that goodness in Liquify and it's a lot more powerful. All issues relating to the formatting have been addressed accordingly and by adopting LSP along with the the powerful Liquify AST parser, formatting is far less error prone and won't mess up your code like it has a habit of doing at times. It is still powered with [PrettyDiff](https://prettydiff.com/) but now relies more on Sparser. Because Liquid is almost always infused with HTML, rulesets and formatting options are now geared towards a structure and default that is more suitable to a Liquid use case, with indented attribute wrapping/ordering and much more. 

#### Digesting Prettier/JS Beautify 

Post Liquify release, the ability to digest prettier/js beautify rules will be made available, this is because PrettyDiff is capable of understanding formatting rulesets supplied by these formatters. This means if you're fanboy or fangirl of those tools, Liquify will respect your presets in those files and work with them to facilitate your beatification taste.

### No more injection grammars

In the current version (2.3) grammar injections are leveraged so as to provide and respect HTML Language features like HTML completions and validation supported in vscode. This comes at a cost which I was unaware of when first implementing them and this has led to some errors and headaches. In Liquify, grammar injections have been boycotted but HTML language features have been preserved, this means that Liquid will not interfere with HTML syntax, instead Liquid and HTML are combined and work with one another. I want to apologize to anyone that suffered at that really terrible implementation of mine, I was young and reckless. The new approach will be tightly coupled with variation specific Liquid and Liquify users will get each variation supplied to them as separate languages, for example Liquify will provide: 

- Liquid Standard
- Liquid Shopify
- Liquid Jekyll
- Liquid Eleventy 

You will need to choose what variation and grammar to use, opposed to the current approach of selecting HTML. This also means a separation of syntax highlighting specific to the variation you're working with, happy days!

### Improved Syntax Highlighting
Liquify has gone a level deeper to bring more expressive syntax highlighting to your editor, you will get even more clear and precise highlighting of code, which allows to you better distinguish Liquid from HTML or other languages, additionally, some great improvements were made to structure of textmate files, which improved rendering speed.

### Freemium Model

Liquify will operate on a freemium license model, which means it will be partially proprietary, but chill fam, before you bring upon the wrath of cancel culture, hear me out. Firstly, I am not going to hustle you and all the features you have in the current version are still going to be available in Liquify for free and will forever stay that way. Additionally, most of the features listed here are available for free too. There will be no annoying distractions, popups or "buy me, buy me" bullshit like we all have suffered at the hands of other tech tooling. The extension will function uninterrupted without licensing and again to reiterate the vast majority of features will be free, however some features if you so happen to want them will require a small one-time licensing cost or if you only need to use those features for a project, then you will be able to lease them on a per month basis.

**Alight, how much?**
Currently, I am still working on the intricacies of this, as of now, licensing will be required for features which pertain to the SaaS variations (ie: Shopify) and/or for some advanced features that pertain to non-SaaS variations – to be clear this still being decided, but please note that individuals and agencies will be subject to different pricing.  In terms of cost, for individuals who wants to lease, it will be around the same price as a decent coffee and for the bandits who want to license up, then you will be looking at around the same cost as paperback book. This will be a one time cost, so no ongoing or yearly commitments, you pay, you get, signed sealed delivered.

**Any chance it won't ship proprietary**
There is still a chance that the tool will ship free with no cost and I really hope that it comes to fruition, fingers crossed!

### Conclusion

I've listed the main features you will expect to see in Liquify and I hope anyone that was curious gets a little more insight. There are a lot more great additions shipping, things like `.liquidrc` intellisense, an optional theme, semantic highlighting, inline comment ignores & formatting, improved support when using Liquid in JavaScript, CSS, SCSS, Markdown, various extensibility, incremental parsing and much more. 

The official release is getting closer. To give an ETA is still hard but users can rest easy knowing that it should conclude and go out well before we close this shitty year of 2020, so in the next few months you can expect to be using Liquify. 
 
Thanks,
Panoply.

**Comment by redareda9** [comment]
I guess the extension isn't maintained anymore?
As no news from Aug 9..

**Comment by panoply** [comment]
Before end of year guys. Sit tight.

**Comment by TheAlemazing** [comment]
Any chance for vim users?

**Comment by panoply** [comment]
I’ll have to look into it. I’m sure VIM supports LSP.

**Comment by panoply** [comment]
For everyone watching or in this thread. 

Currently in the process of the final steps for Liquify. I'm likely to get the beta out in the next couple of days, this initial release is going to include the vast majority of all bug fixes and ship a bunch of new features etc. It's going to be a breaking change, so for those who are working with the extension on a everyday basis, do expect upon updating that you will need to make some manual adjustments (which I will document).

Thanks all for your patience. The wait is nearly over!

**Comment by ff6347** [comment]
@panoply Thank you for working on this. Can't wait to see what is in the 🎁

**Comment by Maxime-cn** [comment]
Just starting out with VSCode so it comes handy I can't wait to see! If that catches your fancy, a youtube tuto could be very helpful to get the setup right first time and showcase the features, how-to etc. (just a suggestion).

**Comment by panoply** [comment]
Thanks for the input @Maxime-cn. I'll try and cover as much as I can via [liquify.dev](https://liquify.dev).

**Comment by nova93** [comment]
![image](https://user-images.githubusercontent.com/3315728/106389416-1cf8ff00-63db-11eb-8a83-1fe580e92a5b.png)
Go go go! Not much time left :-D

**Comment by wisniewski94** [comment]
@panoply bump :)

**Comment by craigjamesdobson** [comment]
Hey @panoply 

Any update on this?

**Comment by samueldsquire** [comment]
👀

**Comment by russellr922** [comment]
@panoply I dropped $50 on this issue on Bountysource(https://www.bountysource.com/issues/92909946-liquify-and-the-future-of-this-extension), I know it isn't much but just looking for a rough timeline.  This is highly desired in the Shopify ecosystem.

**Comment by HugoHeneault** [comment]
Just discovered liquify -- seems awesome! Hope you're going well @panoply and that we can see your work soon. :-)

**Comment by Maxime-cn** [comment]
Hi @everyone,

Since the topic is ignored and liquify.dev website no longer exist, I assume we can no longer expect a new release any time soon, hopefully all is good for @panoply. 
What is still the best way to use vscode with Shopify development nowadays? If someone can share a list of the best reliable extensions he uses, along with its config, that would be helpful for a lot of us, because it seems this topic is not tackled very well anywhere on the internet, and this issue would at least find some answers worth exploring for the time being.

**Comment by panoply** [comment]
Hi all, 

I am sorry I have not been able to release this as yet. I have been going through some really hard times this year. Without going into to much detail but to give you all some answers. Right before concluding and preparing to launch, I went through a pretty heavy breakup with my gf of many years. I lost my house, my dogs and I ended up having to leave country I was living in. 

Please bare with me. I will have it ready in due time.

**Comment by wisniewski94** [comment]
@panoply this is super bad, if there is anything I can do in order to help you then please DM me on Twitter or send an email 🤙 This is what the open-source community does - helps each other.

Meanwhile, I think we can put this on hold, there are more important things :)

Stay strong mate, thanks for letting us know 💪

**Comment by russellr922** [comment]
Man, that’s a lot for anyone to take on. Hope you are doing okay. Thank you for the taking the time to give us an update.

**Comment by TheAlemazing** [comment]
Wow, that's a big hit. Stay strong mate. Thanks for the update 🍀

**Comment by Matb85** [comment]
Huh, hope you'll make ends meet.
Hang in there!

**Comment by rickydazla** [comment]
GF, house, country... but your dogs dawg?! 😩

**Comment by panoply** [comment]
Thank you all for you kind messages and patience. It means a lot to me, truly. I began working on the project again today, trying to pick-up where I left off on it and get the last of stuff ready. I will aim to get a beta version out as soon as I can for you all. Once I can get the initial beta out, I will provide an invite to a discord which is mainly for Shopify Developers but we have a Liquify channel laying dormant so if any issues occur in the beta you can just shoot a direct message.

---

**Side note:** The house, breakup and country were not the hardest part, I am used to moving around but not having my dogs has really messed with me, I guess its true that _hell hath no fury like a woman scorned_. I like to think that my pampered little hairless terrorist will be with me again one day, the other 2 I miss dearly too but they were my ex's more than they were mine and at least they are all still healthy (even if they don't get the big walks that they grew accustomed to with me). They say hope will drive a man crazy, but without hope, we haven't got much. Cherish your dogs folks as they are biggest gift in the world to us humans. Also, for what it's worth, no matter what the circumstance, always make sure you've got equal legal standing in relationships, cause love only goes so far in some cases 😞. 


<img src="https://user-images.githubusercontent.com/7041324/117554348-6058ae80-b057-11eb-9905-8cc246123947.jpeg" width="20%" />

**Comment by samueldsquire** [comment]
All the best @panoply - Stay strong bro 👊

**Comment by panoply** [comment]
I just wanted to keep you all updated. I will be accepting beta testers soon. Below is a couple of screen captures the showcase some the features:

Liquify is smart enough to know exactly how its grammar should be expressed. It knows what is acceptable, For example, below we are validating the `replace` filter and as you would with any linter tool, we get live feedback as we type: 

https://user-images.githubusercontent.com/7041324/119270014-dd108e80-bbfa-11eb-8bee-03c20858319b.mov

In addition, we have hover descriptives. All Liquid tags, filters, parameters, objects and object properties will provide hovers which reference the documentation pertaining to the variation (Shopify, Jekyll or Standard) we are working within. As you can see, HTML Intellisense features are no excluded, instead Liquid works together with HTML features. 

https://user-images.githubusercontent.com/7041324/119270094-527c5f00-bbfb-11eb-934f-6f9fe10db8b3.mov

Here we validating how an object should be expressed:

https://user-images.githubusercontent.com/7041324/119270126-78096880-bbfb-11eb-948b-972761e0e1d9.mov

Here we are validating Shopify section schema. Notice the validations, completions and formatting all work in sync. This feature essentially treats the contents of `{% schema %}` basic tags in the Shopify Liquid variation as its own JSON Schema store. The data entered here is also cached and provided in object completions within the files. 

https://user-images.githubusercontent.com/7041324/119270190-cd457a00-bbfb-11eb-9d98-3cf84900c722.mp4

Here we have Object property completions:

https://user-images.githubusercontent.com/7041324/119270264-26ada900-bbfc-11eb-8789-711028349471.mp4


There are a few more aspects to conclude with, but it's safe to say we at the home stretch and this nearly year long endeavor will soon make it to the big time! Again, thank you all for the patience and understanding. It's been very difficult to get this right and with everything going on in my life, made things a little harder. I will update you all as time progresses.

**Comment by samueldsquire** [comment]
This is going to change the game 🙌

**Comment by panoply** [comment]
Current update:

**Parser and Scanner** 
This part of the project is at 80% completion. The parser supports incremental updates and can walk a 15k line document in less than 400ms. It still needs some work as the scanner does a lot of the heavy lifting, handling validations, tracking cursor position etc etc. Because the parser is working alongside the vscode HTML Language feature, optimising this has taken a long time, because I did not want it to interfere with the vscode implementation. It's operating well at the moment. 

**Completions** 
Completions have been rather difficult to get right. So many developers are accustomed to leveraging snippets and I wanted to move away from snippets in Liquify. Instead, the approach is similar to that of JavaScript intellisense. Trigger characters are wired up in Liquify, as I have before mentioned, I will not ship snippets in Liquify in favor of completion. I wanted to ensure a smooth adaption for developers, it  was really important for me. The result was the following:

In Liquify, tag completions appear when you input certain characters at certain points. For example sake, I will describe the the percent character. When you type `%` all tag completions will show (not objects, tags). I have set a whitespace trigger threshold for tags, filters and output objects (`{{`). This means they will still be available when you type `% ` (for example) and you can cancel out of the completions by hitting escape. The percent character and curly brace is not apart of HTML grammar syntax, so the likelihood of those using the Liquid template language to reach for such characters outside of language itself is minimal (unless you are infusing Liquid with JavaScript, which you should avoid as much as possible), but I digress.  

When you select a tag completion, the remainder of the tag is generated. Each step will provide additional completions, for example:

`%` - List of tags is displayed
`% ` - List of tags displayed is persisted for 1 whitespace character (optional, you can also run completion without space)
`% if` - The `if` tag is asserted (typed) and at this point you either hit `return` or `tab`. 

Liquify generates the tag. For example sake, your cursor is represented as `^`

`{% if^ %}  {% endif %}`

Your cursor will be positioned after the `if` keyword and validation will notify you that a condition is required. When you begin to enter your expression, variables you have defined will be listed first. In Addition, object values will complete at this point too. Let's say you have defined your condition:

 `{% if foo^ %} {% endif %}`
 
 The cursor will move to the end of the `foo`keyword, if you were to enter a `.` or `[` character, Liquify will check for any objects names that have properties under `foo` and they will show up if available. Let's say we do have an object in the Liquid variation called foo that has a property value of `bar`:
 
 `{% if foo.bar^ %} {% endif %}` 
 
 At this point, you will be able to `tab`out which will move your cursor to the inner content of the condition block:
 
 `{% if foo.bar %} ^ {% endif %}` 
 
 However, if you wish to express a comparison against that value, by typing space, the operator value completions are listed: 
 
 `{% if foo.bar ==^ %}  {% endif %}` 
 
 - !=
 - or
 - <
 - >
 - and
 
 > Depending on the Liquid variation you are working with, other operators will show, for example in Shopify you get the additional `contains.
 
 If you do not express the comparison, Liquify will complain that you have incorrect syntax. This entire development flow is used for objects, tags, filters and tag contexts. Hitting `cmd+s` will auto-format and for those who like to leverage whitespace trims, a formatting rule is available that will automatically apply them to tags, you can control this as you would any linting based rule. 
 
 Completions are triggered when the following characters are typed:
 
 - `{`
 - `%`
 - `|`
 - `:`
 - `[`
 - `"`
 - `'`
 - `,`
 
Depending on where you are entering these characters different completions are loaded, for example you cannot trigger `%` tag completions within tags and you cannot trigger `|` filter completions in tags that do not accept filters. The linter and validation will do their best to notify you when you have a made mistake. The most important aspect and take away is that stepping back from snippets in favor of trigger completions might feel extraneous at first, but I can assure you, it is a much faster and more efficient way to write code (regardless of the language). In addition, it is the best way to trigger valid completions for specific Liquid variaitons.
 
 **Formatting**
 
 This aspects is done, there are a bunch of rules and formatting styles you can customize to your own preferences. The formatter leverages the powerful Sparser parser and PrettyDiff engine. These are rather older but suffice really nicely with Liquid. In addition. I have provided a schema store so you guys can understand what each rule does. Below is some of the formatting rulesets that will be shipped:
 
 
 ```json
 {
  "engine": "shopify",
  "format": {
    "ignore": [
      "ignore/*.liquid"
    ],
    "liquid":{
      "dash_trim": "off",
      "force_filter": false,
      "force_line": false,
      "quote_convert": "none"
    },
    "html": {
     "correct": true,
      "force_attribute": false,
      "preserve": 1,
      "wrap": 80,
      "preserve_text": true,
      "space_close": true
    },
    "javascript": {
      "preserve": 1,
      "method_chain": 0,
      "quote_convert": "none",
      "format_array": "indent",
      "format_object": "indent",
      "braces": false,
      "no_semicolon": true
    },
    "scss": {
      "css_insert_lines": true,
      "preserve": 2,
      "braces": false
    },
    "css": {
      "css_insert_lines": true,
      "preserve": 2,
      "braces": false
    },
    "json": {
      "preserve": 1,
      "braces": false,
      "brace_style": "expand",
      "brace_line": false,
      "format_array": "indent",
      "format_object": "indent",
      "tags": [
        {
          "tag": "script",
          "attr": "application\/json",
          "type": "html"
        },
        {
          "tag": "script",
          "attr": "application\/json\\+ld",
          "type": "html"
        }
      ]
    }
  }
}

 ```
 
 Notice there is a `tags` option for each embedded language. You can set Liquid or HTML tags to format in a customized manner. For example, you might capture some JSON in a capture assignment variable, eg: 
 
 ```
 {% capture some_json %}
 {
   "foo": "bar"
 }
 {% endcapture  %}
 ```
 
 There is nothing stopping you from informating the formatter to beautify the contents of the tag in JSON format, for example sake you could achieve this by defining the tag in the `json` ruleset, eg: 
 
 ```json
 {
    "json": {
      "preserve": 1,
      "braces": false,
      "brace_style": "expand",
      "brace_line": false,
      "format_array": "indent",
      "format_object": "indent",
      "tags": [
        {
          "tag": "capture",
          "attr": "some_json",
          "type": "liquid"
        }
      ]
    }
}
 ```
 
Liquify will then format any capture tag attributed with `some_json` according the the rulesets you have defined in the JSON format rule block. This feature is pretty cool, it prevents the formatter for running wild, give you some sanity. The `attr` field accepts a Regular expression string format. 

> Note that the `tags` ruleset will only be applied to markup (HTML) tags or Liquid tags.

----

Again, thank you all for your patience. I would be very much interested to here thoughts about completions aspect. We are close guys. It's just such a big build and I have had to learn so much. I really do not want to release it without it being stable. Thanks for keeping with me on this and I will share some screencasts of the completions feature this week. 

Stay happy and stay healthy.

**Comment by panoply** [comment]
### Update

Below is the final ship list for v1.0.0-alpha.1 which will be available this month (yes, this month no bullshit). Not all features will make it into the initial rollout and those who wish to opt-in for the alpha will need to report on what issues they encounter before we can move to a beta release and then to a pre-release (pre-release will be available on marketplace). All those who wish to install the v1.0.0-alpha.1 candidate will need to do so manually (not via marketplace) and I will inform on the steps for that in the new Liquify repository I will make public.

> I recommend that those of you who are new to development or do not feel completely confident in the nexus refrain from installing and using v1.0.0-alpha.1 candidate when it become available. Those of you are veterans of the game or have good standing knowledge please help me in reporting back on what issues you encounter. 

### Features

- Syntax Highlighting
- Completions
- Validations
- Formatting
- Codelens
- Hover Descriptions

### Syntax Highlighting
As mentioned previously, Liquify does not inject grammars into HTML unlike the current version. Instead grammars will be provided on a variation specific basis which means that files using different extensions will either attempt to automatically assign the grammars to the language and set the mode or require you to manually select the language mode. Below are the languages (grammars) shipping in this candidate:

- Liquid (Shopify) - Shopify Liquid
- Liquid (Jekyll) - Jekyll Liquid
- Liquid (11ty) - Eleventy Liquid

### Completions
The release will include completions and hover support for all tags and filters which belong to the [standard](https://shopify.github.io/liquid) liquid variation. Unfortunately, Jekyll and 11ty variations for completion support will be a post-release feature. The Shopify Liquid variation (for the most part) has been mapped and thus completions will be available for almost all tags, object and filters. There are likely to be issues which I need reporting back on, below will be things to look out for:

- Invalid or incomplete descriptions.
- Scope issues (ie: certain objects made available where they otherwise should not be available).
- Variable Tracking (ie: Render tags assigned variable via parameters not showing)
- Missing object properties
- Expression completions not showing (ie: `{{ object['prop'] }}` where `prop`completion does not show)

These aspects are just minor things and easily amended in the code base, but I will need reporting on where and how the occurred and I will inform on how to report once the candidate becomes available. 

### Validations

This is likely to be the most challenging and errors might show when its valid syntax or code. Not all tags will have validations implemented on them, but the vast majority will. If a validation appears then you will need to inform on where and how it occurred and also be sure that it is an defect in the parser and not just shitty expressed liquid. The validations (linting) capability works different to how other linters you might be using operate and in this candidate we are going to be getting the basics right before exposing rule sets to be applied and enforced on the template language. 

### Formatting

Because the formatting feature is using PrettyDiff and the Sparser parser we have a long list of format options to work with that are otherwise not available in Prettier or JSBeautify. Those installing the candidate will need to inform on where and how it occurred, also be sure that the formatting is not adhering to the options you have set. In the `.liquidrc` file all the format options are using a schema store which inform on what the rule does to syntax, for example:

https://user-images.githubusercontent.com/7041324/121796213-b8b33b00-cc17-11eb-9a5f-058fbad01333.mov

If you feel that the descriptives do not do a good job of describing the option, you should inform. I will ship the candidate with a basic (recommended) ruleset pre-defined. I am yet to come across any major issues with the formatting aspect, like those seen in the current version, however I am not testing against all edge cases. 

### Liquify Project

Liquify is written mostly in TypeScript (some packages are JS) and exists a monorepo built atop of the powerful [pnpm](https://pnpm.js.org). The repository will be released partially open-source and partially closed. It consists of multiple packages which one could consume on the NPM registry at will work outside of Liquify (if one so happens to require them), the main packages include:

- @liquify/liquid-parser
- @liquify/liquid-language-specs
- @liquify/liquid-schema-stores
- @liquify/liquid-language-server

In addition, there are multiple smaller packages used in the Liquify development process, these are mostly internal (project-specific) and/or Rollup plugins which help hook into the transpilation process when compiling the distributed files. The monorepo is currently private but I will make public in the alpha release. I am informing upon this now so anyone who wishes to contribute can get some idea of that to expect on the source-side. Importantly, some packages will be excluded from the public facing repo.

---

That should be about it, if you do read these updates and are invested in the progress or would like to hear, please let me know and we can connect. It is difficult to work on this solo and I do not have any friends within the development community as my day job is within the Fashion industry (I'm a rogue dev) so bouncing ideas is really appreciate and beneficial. 

Thanks and the next update will contain the visx.

**Comment by samueldsquire** [comment]
Exciting times!

**Comment by tr1s** [comment]
Is this out yet officially? I see available June 2021. Mostly commenting to keep track of thread.

**Comment by panoply** [comment]
I had some delays because of my day time job. It's mostly ready for beta release and I am testing everything, little hesitant to drop the beta. Some last minute features I cooked in were HTML parsing and linting capabilities. I wanted to actually bring intellisense support for embedded JavaScript (`<script>`) regions but this is not as easy as it may seem which is actually the main thing holding up the beta release as I really wanted this cooked in before giving users the the first release. I managed to get intellisense capabilities for the following embedded regions:

- `{% schema %}`
- `{% style %}` 
- `{% stylesheet %}` 
- `{% stylesheet 'scss' %}` 
- `<style>`

So you will get completions, hovers, formatting and validations for code written in those regions. but `<script>` and `{% javascript %}` regions only have syntax highlighting and no completion or intellisense features, which is rather annoying. 

I am just not fully comfortable dropping the beta just yet but it could be any day now that it become available. Below is a screen which gives a preview of several of the features available.

https://user-images.githubusercontent.com/7041324/126675458-fa52bdbf-7fea-49b7-8a89-279d6b67d44f.mov

**Comment by panoply** [comment]
Just to add:

I have not done a great job of documentation, so this is taking some time. Because there are going to be a lot of Jekyll users or those using other variations that are not Shopify Liquid, ensuring users know how to extend the extension to support custom tags and in addition how to customize the linting features is also another delay issue. Navigating the code base will take seasoned devs at least a day or two cause its a rather large project so getting help from others in the community is not really possible, it sort of just all rests on my shoulders atm.

There are 3 main aspects to the tool, which are:

1. Parser
2. Specifications
3. Language Server

The parser does a lot of the heavy lifting, the Specifications are used by the parser and are hard coded data files, similar to Custom Data files some of you might be aware of with the HTML Language Service that is maintained by the vscode team. A similar approach has been adopted to Liquify. This allows developers to extend and maintain custom tags, filters and objects in Liquid variations that are not supported natively in Liquify. I have done a terrible job in documenting how one can extend, contribute or modify those, so by releasing it at this point without having concluded that aspect would possibly come back to haunt me, as example here is the Shopify variations `{% paginate %}`specification reference:

<img width="1015" alt="Screenshot 2021-07-22 at 19 32 55" src="https://user-images.githubusercontent.com/7041324/126683235-91c4ed44-130f-41f3-b324-ef31b0248d04.png">

One can actually implement a completely new tag, following that same schema directly within a `.liquidrc` file but I have not informed on how to do it, so essentially this awesome feature that Jekyll or even 11ty users will no doubt use and require will have no information on how to do so. For the sake of clarity, here is how one would actually do this.

Lets create a new Liquid tag called `marvelous` which accepts some arguments, those arguments are of different types, the first argument is a string and its **required**, it must have a value of `marvin`. The `marvelous` tag also accepts an optional 2nd argument within its tag that can be the value of `hagler` - so this custom Liquid tag could look like this:

```js
{% marvelous  'marvin' %}

// or

{% marvelous  'marvin', 'hagler' %}
```

The way in which one would describe this custom tag and in addition get validation and completion support in their editor would be to provide it as a specification reference as aforementioned in the `.liquidrc` file, it would be done like this:

<img width="732" alt="Screenshot 2021-07-22 at 19 45 40" src="https://user-images.githubusercontent.com/7041324/126684838-b6ab6b6c-c745-4348-86e4-8898cf2a1db4.png">

I mean, such a feature is crucial to make known but I just have not documented how to do this in a good way, this barely touches the surface and is the most basic example case. One could go way further, inform that a tag requires a parameter or accepts an object etc etc.

---

The documentation is difficult tbh cause aside from the specifications element of the project, the linting aspect also needs context around what each rule does. For example, one ruleset is `syntactic_placements` which essentially validates the placement of either HTML or Liquid tag pairs, eg: 

https://user-images.githubusercontent.com/7041324/126678681-be98d0c1-6441-49a6-8f97-07b095f58352.mov

For most, this is easily to understand, but the novice developer who installs the beta would potentially get a bunch of validations for issues like this and not understand what it means and just assume something is wrong with the extension and not their code so documenting this in a non-technical way i've found to be super difficult. Another example of this would be the `extraneous_whitespace` rule which is an auto-fix ruleset that will trim the inner spacing of Liquid tags, eg:

https://user-images.githubusercontent.com/7041324/126679372-cbbae237-1189-4ed4-8013-b364fe256c77.mov

A lot of users will auto-format, so such a ruleset will be a no-brainer but a lot of users will not auto-format or auto-fix lint issues and may not care about extraneous whitespace within tags, but the problem is I have not documented or described the ruleset and it can only be turned off within the `.liquidrc` config file. 

When it comes to the HTML linting aspects, because I cooked HTML parsing into the core, we are able to validate HTML syntax rather efficiently, in addition get a ton of cool features. It's not uncommon for developers to forget how tags are placed, so sometimes you will get code like this:

https://user-images.githubusercontent.com/7041324/126680092-476eeac5-094e-4e6e-a55c-347d3c625e6d.mov

What you see there is a `</div>` tag incorrectly placed, where the `{% endif %}` tag is located, that `</div>`  should be before the `{% endif %}` - there are multiple ways you configure how Liquify notifies of such a code issue, because an error like this will happily pass through a Liquid engine but will wreak havoc on the HTML side, the defaults I shipping with the extension does not do the best job of informing about what has gone wrong, again my lack of documentation prevents users from understanding how to handle these aspects in their own efficient way.

Staying on the HTML aspect, users have the ability the validate attributes, but this might not always be desirable, this is default ruleset, wherein a user will get completions for HTML tag attributes as they type but the moment they type an attribute that is not valid or apart of the HTML official spec, you will get validation warnings, eg:

https://user-images.githubusercontent.com/7041324/126681534-2deee485-349f-4f23-bce2-a18a7c038596.mov

This ruleset can be turned off, but again I have done a terrible job of documenting it, in addition some users may not want completions to activate on HTML attributes, such I have not yet made available as an option. The beta release can go out but I am just highly hesitant because of issues, which are not really issues but just aspects that are not clearly informed upon not being completed or accessible to others. 

The last aspects I wanted to touch on in this update is the new Shopify online store 2.0 features. It would be crazy to not ship support for this, luckily everything is covered but ensuring that Liquify has capable context of everything put me back a week or so.

Hope this gives everyone some answers.

**TLDR:** I have done shitty job of documentation and informing upon options and feature which is why I have not released the beta, I need to finish this before making it available. Sorry.

**Comment by wisniewski94** [comment]
@panoply it looks great! can't wait for beta :)

**Comment by jakew009** [comment]
@panoply this is probably going to sound like a weird comment but please make sure you find a way of turning this into a business. You have put a colossal amount of work into this and it sounds like you have had a pretty **** year, so deserve some luck and not to be stressing about the day job :)

**Comment by loicnestler** [comment]
Hey @panoply!
First of all: thank you for your work so far! I've read this thread and am amazed of the amount of effort you put into this!

I was just checking by to see what's the progress on the closed beta so far? 
I've accepted a job for a client of mine who want a custom shopify theme and I haven't found a nice setup for VSCode yet to work with all features you're planning to implement (formatting, linting, snippets, ...).

In case beta isn't public yet: does anyone of you guys know another (at least partially) prod-ready setup / extension bundle for a pleasant workflow in VSCode while working on shopify themes?

Thank you!

**Comment by andronocean** [comment]
This looks beyond cool! I just stumbled across this thread while trying to figure out how to get functional syntax highlighting for liquid inside `<style>` tags (and get stylelint to shut up about syntax errors). Can't wait to try it out — docs or no docs 🤠 If there's beta space available, I'm interested.

BTW, I'd absolutely buy a paid license if the price was right — this would be a huge help for me. I mostly work with Kajabi themes that support _some_ of Shopify's syntax, plus some of their own, so being able to define custom tags and filters would rock my world.

**Comment by SmartKoolAid** [comment]
so what exactly does liquid do currently? Are there really 0 workflows for this language? How is there a billion dollar industry in shopify and no syntax highlighting/formatting for .liquid files? im quite stunned really lol.

**Comment by iBug** [comment]
Happy new year @panoply and everyone! All the best wishes and hope life is getting better and better for you, so that you can ship your amazing product sooner.

**Comment by jdunham2** [comment]
Just wanted to check if there is any news on this?

**Comment by symonxdd** [comment]
Probably not...

**Comment by panoply** [comment]
I am here guys but I am still going through a lot of stuff in my personal life. It will arrive eventually but I just cannot get to it right now and though in our last communications I was very close to the beta I unfortunately had to stop working on it abruptly cause of my daytime job. 

@toklok has worked on the project and also poured hours into it. There are just so many moving parts to Liquify. I had to take over [PrettyDiff](https://github.com/prettydiff/prettydiff/issues/634) and essentially refactor that project since Austin has moved onto Shared File Systems among other things. 

If someone wanted to join us/me in completing the final bits to get it a beta out, I would be happy to bring you in to the existing code base. If you are familiar with the following then please ping me:

- Language parser design and/or writing algorithms.
- LSP (Language Server Protocol)
- TypeScript
- Understands the VSCode extension ecosystem.

I would OS the Liquify project but I have been hesitant of this for many reasons. Shopify engineers themselves have been unable to provide 1/10th of the capabilities Liquify already supports with their maintained solution and Language Server so even in its current state the code base holds value and I would like it private for the time being.

I am sorry folks. I really am. I have not forgotten or abandoned this. I think about it a lot and feel guilty that I have had people wait for so long.

**Comment by panoply** [comment]
**Updates:**

I know I sound like a broken record but we are closer. I have decided to strip a lot of features and instead focus on getting a workable version out with just the fundamentals, bug fixes and the most useful aspects (completions, formatting, highlights configs etc). In addition to Liquify, I have also been working on [Syncify](https://github.com/panoply/syncify/tree/next) which is a theme-kit alternative (more on that below). I will give you an update on each aspect and some insight on the current status quo.  

## Liquify

As previous comments over the last few years I have shared with all those interested, the project has been a large one to take on individually. I have had a rough few years, together with relationship issues and just personal stuff I was going through it was difficult to see the light at the end of the tunnel so-to-speak. I apologize for the continually broken promises. Liquify as aforementioned couples multiple projects together and is developed within a multi/mono-repo environment. Within the project environment exists several essential other projects that folks will be able to independently use. I have been over the Specifications, Parser and Language Server in https://github.com/panoply/vscode-liquid/issues/56#issuecomment-885112963. Below are the **key features** of Liquify:

- Formatting Support for Liquid in HTML, JavaScript and CSS
- Completion Support for Shopify API
- Advanced Syntax highlighting.
- Variation specific support, ie: Standard, Shopify, Jekyll etc
- Validation/Linting support for Tags, Filters and Objects

Keep in mind that when I say I have stripped features I am referring to different areas of these capabilities Liquify provides. For example, completions will not support `{% schema %}` section defined data references in the release. Below I will go over some additional key projects within Liquify, some of which I have already open sourced and folks can track the progress.

## Prettify

The project [Prettify](https://github.com/panoply/prettify) is where those curious can see the brunt of the work for the formatting support that Liquify will operate atop of. Prettify is a hard-forked variation of the wonderful [PrettyDiff](https://github.com/panoply/prettydiff) project which I have actually taken over. The original author [Austin](https://github.com/prettydiff) has since moved onto [Shared File Systems](https://github.com/prettydiff/share-file-systems) and given PrettyDiff is an extremely powerful formatting/beautification tool while working on Liquify of the last couple of years it become apparent that stripping and refactoring the original project was the best decision moving forward and given the familiarity with the PrettyDiff project I already I have, providing additional maintenance was something I wanted to offer.  

[Prettify](https://github.com/panoply/prettify) will be the Liquify specific formatter that Liquify uses under the hood but it will also be available to those who want to leverage it in other projects or even online. Prettify will be ready to consume and use upon the release of Liquify. 


## Syncify

Thought most of the companies I work with or have worked with have since moved to [Centra](https://centra.com/) from Shopify. I still do work with a couple of clients who use Shopify and wanted to create a powerful theme-kit equivalent. I been putting together [Syncify](https://github.com/panoply/syncify/tree/next) which while still a **WIP** is something that will be published together with Liquify. Syncify will be made available independent of Liquify but Liquify will provide built-in support for projects who decide to use it. It is geared towards Shopify Theme developers and will enable you to elegantly compose and develop themes. 

Syncify will not impede on developer preferences and tool appropriation when it comes to handling asset files, so using tools like ESBuild, Webpack, Rollup, SASS etc etc to process asset specific files is made possible with ease and its main focus is to construct theme structures. Below is some the **key features** Syncify provides:

- Upload, download and watch multiple storefronts and/or themes.
- Clear, concise and informative CLI logging.
- File sync support, upload and download to and from store files endpoint.
- Supports HTML + Liquid and JSON minification.
- An elegant directory based metafields sync approach using JSON files.
- Built-in support for SCSS and CSS transpilation using SASS Dart and [PostCSS](https://postcss.org/).
- SVG Sprite and inline SVG snippet generation using [SVGO](https://github.com/svg/svgo).
- Intelligent path mapping capabilities for custom theme directory structures.
- Digests existing build tool configurations for asset transformations.
- Prompt based CLI/TUI and exposed module API for script usage.

The current readme is not upto date, but goes into more detail. Please keep in mind that [Syncify](https://github.com/panoply/syncify/tree/next) will replace [shopify-sync](https://github.com/panoply/syncify) and is not yet available on NPM registry. 

---

For those who are interested getting their hands on the BETA release of Liquify, please shoot me an [email](n.savvidis@gmx.com) and I will arrange the VSIX when ready to you. 

Thanks for everyone's patience, again, sorry for the broken promises and delays, I won't let you down guys.

**Comment by panoply** [comment]
Fixed!

**Comment by loicnestler** [comment]
@panoply invite is sadly expired :/
can you provide a new one?

**Comment by MattBudz** [comment]
what's the latest new on this?

**Comment by panoply** [comment]
## Update (Liquify is now Open Source!)

Just wanted to quickly update folks in this thread with ongoing development and supersede coming for Liquify. The project is now public and open source and can be found [here](https://github.com/panoply/liquify). Though a lot of the code is still yet to be pushed from my local machine, curious developers can browse the monorepo and get a glance of the algorithm, parse approach and overall architecture that has gone into Liquify. 

Those who are using the current version of this extension have likely been enjoying some of the new capabilities which I've made available in the lead up to Liquify. The next version **3.4.0** will bring even more great features to developers and it will also be the last minor version to be released as following **3.4.0** the Liquify supersede will become a reality.

## What to expect?

Liquify will be initially made available under a preview release in the vscode marketplace. This will essentially be the "beta" and users will be able to install and use it while also reporting any bugs they encounter. Previous comments in this thread have touched on several key features, most of which are already available in the current version, but I will outline the additional capabilities you will get in Liquify.

### Validations

Liquify supports complex validating of Liquid and HTML. You can control the level of validating which is incurred, similar to how you control linting tools (eslint, stylelint etc). Validations are briefly outlined in https://github.com/panoply/vscode-liquid/issues/56#issuecomment-885112963

### Types

In Liquid, Type support is somewhat unheard of but Liquify solves that. Essentially, typed assertion is available, which means you'll have validations and tips when type issues occur (eg: you try to iterate over a number type). 

### Definitions

Take the following code sample:

```liquid
{% render 'some-file', arg_1: some.object, arg_2: 'some string' %}
```

In Liquify, it understands imports of this logic, so when you access the `arg_1` or `arg_2` values passed to the file, it will provide you with completions but also definition scopes, this means that you will be able to `cmd+click` on the reference and it will open the relative document import. The definition capability goes beyond just renders, but also works applied to inline code, for example:

```liquid
{% capture foo %}
  Hello World
{% endcapture %}

{% # 200 lines later %}

{{ foo }}
```
Whenever you hover your cursor over `{{ foo }}` then you will have context and code preview of the capture contents, including range reference so you will be able to `cmd+click` and be take to the exact point of location in the document where the capture occurred. 

### Linked Documents

Similar to definitions, linked documents capability (which will actually be available in **3.4.0**) provides you with project aware linking of different doucment. For example, take the following:

https://user-images.githubusercontent.com/7041324/209701035-75f2a95a-e5a7-4f8b-84fb-3544ed0a3216.mov

This will be available for all variations, not just Shopify but also Jekyll, 11ty etc. 

### Completions

The current version of the extension already supports an extensive list of completions. The Shopify variation already has support for tags, objects and filters. In version **3.4.0** I will actually extend this to locales, and `settings_data.json` files for Shopify, eg:

https://user-images.githubusercontent.com/7041324/209701523-1d6001d7-f951-4347-bd52-9d57ae3ff235.mov

In Liquify, this will be taken to new levels. Not only will all the current completion capabilities be available, you will also have the ability to pull down Shopify metafield from remote stores and have completions exposed in the editor. In addition to the metafield completions, assets files are also available. 

Lastly, assignment based completions will also be supported and various other Jekyll/11ty geared completion logics alongside completions for filter and tag arguments.

### Signatures

Tag and Filter based signature support is exposed in Liquify. This will enable you to also have contextual awareness of the a tag or filters expected syntax structure and type formations. Very cool feature. 

## Shopify Liquid vs Liquify?

Shopify has recently began working towards a solution which intends to provide similar capabilities of Liquify. Their solution provides similar features but what is offered is less than that available in version **3.2.0**. The features shipping in Liquify and the next release (#138) intend to solidify this extension as the preferred developer tooling solution for Liquid, across the text editor nexus. 

My goals are to provide a better product, one which does far more and extends support cross platform. Users who have been enjoying the features since bumping to **v3.0.0** have got glimpse of what is to come and I hope that the current list of available features have helped in your productively. 

## What's next?

Following the **3.4.0** all energy and work will move toward the Liquify supersede and all additional threads, updates and progress can be tracked within the [Liquify](https://github.com/panoply/liquify) monorepo repository. Given that the Liquify has been open sourced, discussions, feature request and other collaborative ideas or work will be handled there. 

Lastly, Liquify is going to be **FREE** and **OPEN SOURCE** with no licensing or costs imposed. My hope is that users choose to use it over the alternative. That will make me truly happy. 

I'm going to close this thread now. Wishing all health, happiness and a good new year! 

See you on Liquify!

**Comment by ajiho** [comment]
I have read the entire article and found that you have devoted a lot of time and energy to it, which is really remarkable. Before this, I always stubbornly believed that only official implementations were the best for everything, but since using your extension, it has changed my mindset  . [Shopify/theme-tools](https://github.com/Shopify/theme-tools/tree/main/packages/vscode-extension) I think the user experience of this official expansion is really poor. I think Shopify should directly use your expansion as an official recommendation and pay you a salary. Today is my first time using your extension, and from the moment I first installed it, I was amazed. I think you have considered the details very well, especially the intelligent prompt that can preserve the original HTML of VSCode in the liquid file . I would definitely recommend this extension to others, and I am planning to use it in my open source project and recommend everyone to install it for a better experience
